File: design-pattern-java/app/src/test/java/design/pattern/java/patterns/creational/singleton/DBConnectionTest.java
package design.pattern.java.patterns.creational.singleton;

import java.util.stream.Stream;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class DBConnectionTest {
  @Test
  public void itShouldCreateADBConnectionWhenCreateIsCalled() {
    String host = "test.com";
    int port = 8443;
    String username = "test";
    String password = "****";
    final DBConnection connection = DBConnection.create(
      host, port, username, password);
    Assertions.assertEquals(host, connection.getHost());    
    Assertions.assertEquals(port, connection.getPort());    
    Assertions.assertEquals(username, connection.getUsername());    
    Assertions.assertEquals(password, connection.getPassword());    
  }
  @Test
  public void itShouldCreateOnlyOneInstanceWhenCreateIsCalledMoreThanOnce() {
    String host = "test.com";
    int port = 8443;
    String username = "test";
    String password = "****";
    final DBConnection connection = DBConnection.create(
      host, port, username, password);
    Stream.generate(() -> Double.valueOf(Math.random() * 100).intValue())
      .limit(10)
      .peek((t) -> System.out.println(t))
      .map(
        (index) -> DBConnection.create(
          host, port, username.concat(index.toString()), password)
        )
      .forEach((testDBConnection) -> 
        Assertions.assertEquals(connection, testDBConnection)
      );
  }  
}

File: design-pattern-jvm/app/src/test/java/design/pattern/kotlin/patterns/creational/singleton/DBConnectionTest.kt
package design.pattern.kotlin.patterns.creational.singleton

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.stream.Stream

class DBConnectionTest {
    @Test
    fun itShouldCreateADBConnectionWhenCreateIsCalled() {
        val host = "test.com"
        val port = 8443
        val username = "test"
        val password = "****"
        val connection = DBConnection.create(
            host, port, username, password
        )
        Assertions.assertEquals(host, connection?.host)
        Assertions.assertEquals(port, connection?.port)
        Assertions.assertEquals(username, connection?.username)
        Assertions.assertEquals(password, connection?.password)
    }

    @Test
    fun itShouldCreateOnlyOneInstanceWhenCreateIsCalledMoreThanOnce() {
        val host = "test.com"
        val port = 8443
        val username = "test"
        val password = "****"
        val connection = DBConnection.create(
            host, port, username, password
        )
        Stream.generate { java.lang.Double.valueOf(Math.random() * 100).toInt() }
            .limit(10)
            .map { index: Int ->
                DBConnection.create(
                    host, port, username + index.toString(), password
                )
            }
            .forEach { testDBConnection: DBConnection? ->
                Assertions.assertEquals(
                    connection,
                    testDBConnection
                )
            }
    }
}